### DISCUSSION TITLE
How, do i make a data select? (seems to Sonic 3)

### DISCUSSION THREAD

**Discussion by The-Enazer** [Techokami/SonicWorldsNext#206] [Q&A] [answered]
Hello, i new in this engine. I wanna make a data select like sonic 3.
But i not how work this engine of sonic.
Please help. lmao (sorry for my bad english)

**Comment by NeonSRB2** [comment, marked as answer]
This engine does not feature saving like that, you have to make that yourself.
Thankfully though, it's extremely easy to do, and the engine already has to save files for settings / controls.
Also, my project is based off of the old 3.5 branch, so apologies for any misnamed stuff here if that's a problem. I will also note that I haven't made the save select myself either, so less free code snippets here.

Before anything, you should make the saving itself. If you want to debug this, try making an extra entry in the level select that, instead of loading a level, loads save slot 1, overriding your character choice if that data is present.

Both saving and loading functions should provide a field that correlates to the name of the save file's actual file. You can either try to concat a number to a hardcoded string, or make the entire input the string name of the save file.

Novv, here's the existing saving function for settings:
```
func save_settings():
	var file = ConfigFile.new()
	# save settings
	file.set_value("Volume","SFX",AudioServer.get_bus_volume_db(AudioServer.get_bus_index("SFX")))
	file.set_value("Volume","Music",AudioServer.get_bus_volume_db(AudioServer.get_bus_index("Music")))
	
	file.set_value("Resolution","Zoom",zoomSize)
	file.set_value("Resolution","FullScreen",((get_window().mode == Window.MODE_EXCLUSIVE_FULLSCREEN) or (get_window().mode == Window.MODE_FULLSCREEN)))
	# save config and close
	file.save("user://Settings.cfg")
```

file.set_value is the important stuff. It's vvhat you're saving! Save PlayerChar1 & 2, emeralds... that stuff.
The stage is the tricky bit though, there's no existing variable for that. Hovvever, PackedScenes (The vvay the engine expects a loadable scene to be stored as) have resource_path variable. You can make a var in Global/Main.gd for the current scene path, and set it to this like so.
The first and second variables can be any string, as long as they are recalled by the same tvvo strings in the load function
```
var scenepath = "this value doesn't really matter for this purpose"
func change_scene(scene = null, fadeOut = "", fadeIn = "", setType = "SetSub", length = 1, storeScene = false, resetData = true):
	scenepath = scene.resource_path
```
Novv by the time you save, you already have the stage you're loaded in (or about to be loaded in).
There vvas a section here about loading the stage but it vvas vvrong.

Supposing you got this all vvorking nicely, novv you got to make the menu, vvhich you just kinda... do. You can do vvhatever you vvant here, so long as you call your load function from Global upon the saves' _ready or something. All that you really need to knovv is that you should be able to make a menu like this godot vvith some yt tutorials or something.

**Reply by NeonSRB2** [reply]
I forgot to mention some stuff; upon the players' _ready function, if there's no checkpoint data, call the save function. The save slot this vvrites to should be a variable that is changed upon loading a file, regardless if the file exists.

Also, you should make the save files call your load function in their _ready funtions, make sure the load function tries to access this file instead of modifying the variables directly, so multiple saves can have different data. vvhen selecting a save file, set the values in the save to Global and load the stage's path.

Oops.

**Reply by Raptorq1** [reply]
@NeonSRB2 that wouldn't probably work he said he wanted to make a sonic 3 save select screen try this instead

https://www.reddit.com/r/godot/comments/16fhfbx/how_would_i_do_this_effect/?rdt=45881

**Reply by r6915ee** [reply]
> @NeonSRB2 that wouldn't probably work he said he wanted to make a sonic 3 save select screen try this instead
> 
> https://www.reddit.com/r/godot/comments/16fhfbx/how_would_i_do_this_effect/?rdt=45881

What NeonSRB2 provided complies with how basic ConfigFile save data systems in Godot work and should function as expected; it is the backend. The menu itself is up to The-Enazer.

**Comment by MrCherry09** [comment]
Did you ever figure this out? I can not get it working for the life of me
